bcipy/tasks/query_selection.py

```diff
@@ -0,0 +1,190 @@
+                query to be selected
+            lam(float): in [0,1] convex combination parameter
+            gam(float): in [0,1] history decay rate for momentum
+            momentum(ndarray[float]): momentum values for each element
+            prob_history(list[list[float]]): probability values for each step
+                in the history for each element in the alphabet
+            last_query(list[str]): final query generated by the system
+        Functions:
+            reset():
+                resets the system memory
+                """
+
+    def __init__(self, len_alp, len_query=4, gam=.9, lam=.9):
+        self.alp = list(range(len_alp))
+        self.lam = lam
+        self.gam = gam
+        self.len_query = len_query
+        self.momentum = np.zeros(len(self.alp))
+        self.prob_history = []
+        self.last_query = []
+
+    def reset(self, lam=.9):
+        """ resets the history related items in the query method
+            Args:
+                lam(float): in [0,1] convex combination parameter """
+        self.lam = lam
+        self.momentum = np.zeros(len(self.alp))
+        self.prob_history = []
+
+    def update_query(self, p):
+        """ with the final belief over the system, updates the querying method
+            and generates len_query most likely queries.
+            Args:
+                p(list[float]): list of probability distribution over the state
+                    estimates
+            Return:
+                pos_query(list[int]): positions of queries in alphabet """
+        tmp = p[:]
+        self.prob_history.append(tmp)
+        self.update_momentum()
+
+        entropy_term = np.array(tmp) * np.log2(tmp + eps) + (
+                1.01 - np.array(tmp)) * (np.log2(1.0 - np.array(tmp) + eps))
+        entropy_term[np.isnan(entropy_term)] = 0
+        # TODO: Stupid magic number
+        reward = (self.lam - 1) * entropy_term + self.lam * self.momentum
+        self.update_lam()
+
+        tmp_alp = [i for i in self.alp]
+        tmp_query = best_selection(tmp_alp, reward, self.len_query)
+
+        self.last_query = [i for i in tmp_query]
+
+        return self.last_query
+
+    def update_momentum(self):
+        """ momentum is updated with the particular probability history of the
+            system.
+            WARNING!: if called twice without a probability update, will update
+            momentum using the same information twice """
+        if len(self.prob_history) >= 2:
+            # only update momentum for previous terms
+            idx_prev_query = [self.alp.index(self.last_query[i]) for i in
+                              range(len(self.last_query))]
+            self.momentum *= self.gam
+
+            for k in idx_prev_query:
+                # momentum = current_mass * mass_displacement
+                self.momentum[k] += self.prob_history[-1][k] * np.power(10,
+                                                                        5) * (
```

Comment:
There's some odd formatting on this line. Are you using a code formatter for this? I switch back and forth between autopep8 and yapf.

Reply:
I fixed the formatting.

I am using auto-pep8 in Pycharm.